persist: chunk up a trace batch into multiple keys

ruchirK opened the pull request:
<!--

Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.

-->

### Motivation

This commit teaches persist to perform trace compaction with bounded memory usage
by:

1. Only ever keeping one BlobTraceBatchPart in memory from each of the two
   batches being compacted at a time.
2. Only keeping one ColumnarRecords worth of merged data in memory at a time.
3. Performing roughly a linear merge on the two trace batch parts and as we read
   data from each trace batch, figuring out the upper bound on data that can be
   compacted, consolidating and merging that, and then placing that in a
   ColumnarRecords to await being written out as an output BlobTraceBatchPart.

(3). is best explained with an example. If we are compacting two trace batches,
A which has two parts, one with keys [0, 10) and one with keys [10, 20) and B has
three parts, with keys [0, 15), [15, 30) and [30, 45) respectively then while
compacting when we observe the first batches from A and B which have keys
[0, 10) and [0, 15) respectively, we cannot merge all of both batches together.

We can only merge the subset of keys in [0, 10), as its possible that subsequent
parts in A have relevant keys in [10, 15).

We don't do the real linear merge as compaction requires us to both forward times
and consolidate multiple updates at the forwarded times. I believe that doing so
is possible in linear time, but doing so in a single pass was complicated enough
that I am leaving it for future work.

```
MZ_PERSIST_RECORD_COUNT=819200 MZ_PERSIST_RECORD_SIZE_BYTES=64 MZ_PERSIST_BATCH_MAX_COUNT=131072
writer/compact_trace/mem/50MiB
                        time:   [215.58 ms 216.61 ms 217.71 ms]
                        thrpt:  [229.66 MiB/s 230.83 MiB/s 231.93 MiB/s]
writer/compact_trace/file/50MiB
                        time:   [450.89 ms 451.40 ms 451.94 ms]
                        thrpt:  [110.64 MiB/s 110.77 MiB/s 110.89 MiB/s]
writer/compact_trace/s3/50MiB
                        time:   [554.41 ms 584.38 ms 621.09 ms]
                        thrpt:  [80.504 MiB/s 85.560 MiB/s 90.186 MiB/s]
```

At the default settings this commit is ~33% slower on the trace compaction benchmark
and roughly identical for file and s3 blob (the s3 blob implementation has really high
variance, and can be anywhere from 800 - 500 ms).

Being slower on mem blobs makes sense, as the new implementation does extra allocations and
deallocations when reading trace batch parts, and has to do more copies into a buffer where
the finalized data sits before it is written out. In absolute terms, we observe a 50ms slowdown
in mem blob which feels significant enough to also present in the file and s3 blob benchmarks.
It's unclear why that doesn't happen.

```
MZ_PERSIST_RECORD_COUNT=819200 MZ_PERSIST_RECORD_SIZE_BYTES=64 MZ_PERSIST_BATCH_MAX_COUNT=409600
writer/compact_trace/mem/50MiB
                        time:   [220.03 ms 225.41 ms 230.34 ms]
                        thrpt:  [217.07 MiB/s 221.82 MiB/s 227.24 MiB/s]
writer/compact_trace/file/50MiB
                        time:   [450.84 ms 451.22 ms 451.69 ms]
                        thrpt:  [110.69 MiB/s 110.81 MiB/s 110.90 MiB/s]
writer/compact_trace/s3/50MiB
                        time:   [564.77 ms 588.75 ms 614.72 ms]
                        thrpt:  [81.337 MiB/s 84.925 MiB/s 88.531 MiB/s]
```

This setting simulates the case where all for a given batch lives in one part. There's
no significant difference in performance between this and the default benchmark settings
which suggests that the overhead of reading and managing state for multiple batches is
reasonable.

```
MZ_PERSIST_RECORD_COUNT=8192000 MZ_PERSIST_RECORD_SIZE_BYTES=64 MZ_PERSIST_BATCH_MAX_COUNT=1310720
writer/compact_trace/mem/500MiB
                        time:   [2.2916 s 2.2986 s 2.3064 s]
                        thrpt:  [216.79 MiB/s 217.52 MiB/s 218.19 MiB/s]
writer/compact_trace/file/500MiB
                        time:   [6.0161 s 6.0258 s 6.0346 s]
                        thrpt:  [82.856 MiB/s 82.976 MiB/s 83.110 MiB/s]
writer/compact_trace/s3/500MiB
                        time:   [3.2980 s 3.5561 s 3.8442 s]
                        thrpt:  [130.07 MiB/s 140.60 MiB/s 151.61 MiB/s]
```

This final benchmark run is the same as the default settings, but with 10x as
much data placed in 10x larger batches, Mem blob exhibits almost perfectly
linear scaling, file blob shows slighly worse than linear scaling, and s3 blobs
actually show a performance improvement in this regime, achieving slightly less
than double the throughput. My unsubstantiated hypothesis is that perhaps at
the smaller sizes s3 blobs are not able to take advantage of splitting a BlobTraceBatchPart
into evenly sized chunks for the s3 uploads (or perhaps there's one large chunk and one tiny
chunk at the end?).

### Tips for reviewer

The first commit has the new compact trace benchmark + a detailed analysis of how they look currently in three different size + num batch configurations, the second commit has the new/modified invariants for trace batches and the third commit actually handles compacting part-by-part with bounded overhead (now without quadratic work :)), and new tests + an evaluation of the compaction microbenchmarks.

<!--

Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.

-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

ruchirK commented:
@danhhz this is a very very rough sketch of what I was thinking

I think the main missing pieces here are:

1. properly doing a merge sort of the chunks
2. Have a new type for `BlobTraceBatchChunk` instead of reusing `BlobTraceBatch` for this purpose
3. Figure out what other metadata I'd want to include in the chunks
4. Obviously, fix bugs :)

danhhz commented on src/persist/src/indexed/background.rs:
```diff
@@ -140,49 +140,105 @@ impl<B: Blob> Maintainer<B> {
-            key: merged_key,
-            format,
+            key: keys,
+            format: formats[0],
```
when we do this for real, I think we'll want to pass the format to set_trace_batch instead so we don't have to deal with this

danhhz commented on src/persist/src/persist.proto:
```diff
@@ -55,7 +55,7 @@ message ProtoUnsealedBatchMeta {
 }
 
```
the desc field of ProtoTraceBatchInline isn't really correct anymore. we could document that it means the associated data is part (but maybe not the entirety) of the batch, but maybe we should just remove the field instead? it was just a nice to have anyway. too bad we won't know ahead of time how many chunks there will be, it'd be nice to store "this is chunk N of M" in the inline metadata. the N part at least is easy

danhhz reviewed (commented):
I'm 95% convinced. I think my major remaining issue is what to do with the inline metadata that we store in parquet

bikeshed: this just reminded me that I wanted to rename BlobMeta, BlobTraceBatch, and BlobUnsealedBatch since the Blob- prefix meant it was something directly serialized and saved as a blob value, which is no longer true for any of them. in light of that, maybe instead of BlobTraceBatch and BlobTraceBatchChunk, we have TraceBatch and TraceBatchBlob (chunk is already used in parquet)? or maybe TraceBatch/TraceBatchPart?

ruchirK commented:
Ok I think a lot of the major concerns are being addressed here:

1. renaming the types / adding the index is happening in https://github.com/MaterializeInc/materialize/pull/10751
2. Changing the set_*_batch calls to take the format as an arg is happening in #10858 
3. I changed this pr to properly do a merge sort, instead of blindly merging chunks together

I think this all needs a lot more testing but I kicked off a long open loop benchmark run to validate that this keeps the memory usage bounded during large compactions which is the whole point here.

ruchirK commented:
This is the memory usage after ~16 hours of running the open loop benchmark

<img width="795" alt="Screen Shot 2022-02-24 at 10 44 19 AM" src="https://user-images.githubusercontent.com/4015012/155557773-5c77c1b3-30e5-4695-988d-455128ceded4.png">

This is a substantial improvement. Previously the memory usage was spiking up to ~all of the data previously ingested which was 40+ GiB

danhhz commented:
\o/

ruchirK commented:
@danhhz this is now ready for a real look! The diff grew quite a bit as I added a lot more validations and tests for multipart trace batches.

Notes/comments for reviewers:

0. I added one change thats not strictly necessary. I disallowed empty BlobTraceBatchParts as now if someone wants an empty trace batch they can express that by having 0 parts. I can extract that into a separate pr if desired.
1. I'm going to squash all of this into one commit. Keeping the structure as-is for the moment to preserve append-only-ness.
2. I'm going to switch to getting all micro/macro benchmark results now, and will update the pr shortly
3. I think my chief concern with the actual feature changes is that its quite a bit more allocation/deallocation heavy than the previous version of the compaction code. I _think_ there's ways to avoid lots of that, but to be honest I have not thought about it too deeply. I'm going to see how affected the benchmarks are to decide whether its ok to punt on that for later.
4. My chief concern with the testing is that we're not actually testing the case where a `ColumnarRecords` fills up in the middle of merging two trace batches, and I'm not sure how to exercise that without sending two very large (~4 GiB?) batches.

danhhz commented on src/persist/src/indexed/background.rs:
```diff
@@ -154,68 +155,160 @@ impl<B: Blob> Maintainer<B> {
+            let mut i = 0;
+            while i < first_updates.len() {
+                if first_updates[i].0 <= threshold_key_val {
+                    let data = first_updates.remove(i);
```
doesn't this do O(n^2) memcpys?

my immediate reaction is maybe VecDeque helps but I haven't thought about it enough to be sure

ruchirK replied:
this is a hell of a catch thank you! I think `split_at` does what we want more efficiently I'll test out all options and add a new microbenchmark

ruchirK replied:
Confirmed - this had terrible performance. The 50MiB compact trace benchmark which takes ~150ms per iteration with `split_at` took ~75 seconds per iteration with the "copy a quadratic number of bytes by repeatedly calling remove" implementation. I didn't bother to analyze that more closely as the single benchmark run took over 12 minutes.

danhhz commented on src/persist/src/indexed/background.rs:
```diff
@@ -154,68 +155,160 @@ impl<B: Blob> Maintainer<B> {
+
+            for ((key, val), time, diff) in updates.iter() {
+                builder.push(((key, val), *time, *diff));
+                if builder.num_filled() > 0 {
```
instead of num_filled/take_filled, why not only take_filled but it returns an Option?

ruchirK replied:
done.

danhhz reviewed (commented):
here's the first few things that jumped out at me. heads up that with the multiwriter prototyping this week, I may not have time to look at this in detail soon

ruchirK commented:
Microbenchmarks are unremarkable as expected

```
ubuntu@ip-10-1-26-184:~/materialize/src/persist$ cargo bench
   Compiling mz-persist v0.0.0 (/home/ubuntu/materialize/src/persist)
    Finished bench [optimized + debuginfo] target(s) in 47.60s
     Running unittests (/home/ubuntu/materialize/target/release/deps/benches-4e83c0fe211f8fde)
Gnuplot not found, using plotters backend
MZ_PERSIST_RECORD_COUNT=819200 MZ_PERSIST_RECORD_SIZE_BYTES=64 MZ_PERSIST_BATCH_MAX_COUNT=131072
Benchmarking end_to_end/replay/file/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 71.1s, or reduce sample count to 10.
Benchmarking end_to_end/replay/file/50MiB: Collecting 100 samples in estimated 71.082 s (100 iterations)
end_to_end/replay/file/50MiB
                        time:   [710.96 ms 712.91 ms 714.83 ms]
                        thrpt:  [69.947 MiB/s 70.135 MiB/s 70.328 MiB/s]
                 change:
                        time:   [+2.4474% +2.8542% +3.2840%] (p = 0.00 < 0.05)
                        thrpt:  [-3.1795% -2.7750% -2.3889%]
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild

snapshot/blob_get/mem/50MiB
                        time:   [21.975 ms 21.987 ms 22.000 ms]
                        thrpt:  [2.2195 GiB/s 2.2208 GiB/s 2.2220 GiB/s]
                 change:
                        time:   [+3.6092% +3.6892% +3.7681%] (p = 0.00 < 0.05)
                        thrpt:  [-3.6312% -3.5579% -3.4835%]
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Benchmarking snapshot/mem/unsealed/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 12.9s, or reduce sample count to 30.
snapshot/mem/unsealed/50MiB
                        time:   [129.29 ms 129.36 ms 129.43 ms]
                        thrpt:  [386.30 MiB/s 386.51 MiB/s 386.71 MiB/s]
                 change:
                        time:   [+1.9052% +2.5162% +3.0956%] (p = 0.00 < 0.05)
                        thrpt:  [-3.0026% -2.4545% -1.8696%]
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
Benchmarking snapshot/mem/trace/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 15.4s, or reduce sample count to 30.
snapshot/mem/trace/50MiB
                        time:   [144.78 ms 144.90 ms 145.02 ms]
                        thrpt:  [344.79 MiB/s 345.07 MiB/s 345.34 MiB/s]
                 change:
                        time:   [+16.703% +17.083% +17.464%] (p = 0.00 < 0.05)
                        thrpt:  [-14.868% -14.590% -14.312%]
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Benchmarking snapshot/file/unsealed/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 13.0s, or reduce sample count to 30.
snapshot/file/unsealed/50MiB
                        time:   [129.76 ms 129.83 ms 129.89 ms]
                        thrpt:  [384.93 MiB/s 385.12 MiB/s 385.31 MiB/s]
                 change:
                        time:   [+3.9966% +4.4009% +4.7622%] (p = 0.00 < 0.05)
                        thrpt:  [-4.5457% -4.2154% -3.8431%]
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
Benchmarking snapshot/file/trace/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 16.1s, or reduce sample count to 30.
snapshot/file/trace/50MiB
                        time:   [146.19 ms 146.31 ms 146.43 ms]
                        thrpt:  [341.46 MiB/s 341.74 MiB/s 342.01 MiB/s]
                 change:
                        time:   [+20.381% +20.493% +20.614%] (p = 0.00 < 0.05)
                        thrpt:  [-17.091% -17.008% -16.930%]
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

writer/log/mem_sync     time:   [66.348 ns 66.600 ns 67.112 ns]
                        change: [-22.366% -8.2379% +8.4224%] (p = 0.34 > 0.05)
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high severe
Benchmarking writer/log/file_sync: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 8.3s, enable flat sampling, or reduce sample count to 50.
writer/log/file_sync    time:   [1.5908 ms 1.6039 ms 1.6171 ms]
                        change: [-2.4068% -1.4233% -0.3947%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

Benchmarking writer/blob_set/mem/50MiB: Warming up for 1.0000 s
Warning: Unable to complete 100 samples in 1.0s. You may wish to increase target time to 1.1s, or reduce sample count to 90.
writer/blob_set/mem/50MiB
                        time:   [17.225 ms 17.377 ms 17.539 ms]
                        thrpt:  [2.7840 GiB/s 2.8099 GiB/s 2.8347 GiB/s]
                 change:
                        time:   [-0.8322% +0.3902% +1.6174%] (p = 0.52 > 0.05)
                        thrpt:  [-1.5917% -0.3886% +0.8392%]
                        No change in performance detected.
Found 21 outliers among 100 measurements (21.00%)
  21 (21.00%) high severe
Benchmarking writer/blob_set/file/50MiB: Warming up for 1.0000 s
Warning: Unable to complete 100 samples in 1.0s. You may wish to increase target time to 26.1s, or reduce sample count to 10.
writer/blob_set/file/50MiB
                        time:   [392.40 ms 392.56 ms 392.71 ms]
                        thrpt:  [127.32 MiB/s 127.37 MiB/s 127.42 MiB/s]
                 change:
                        time:   [+0.0285% +0.0762% +0.1226%] (p = 0.00 < 0.05)
                        thrpt:  [-0.1224% -0.0761% -0.0285%]
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low severe
  1 (1.00%) high mild
  2 (2.00%) high severe

writer/encode_batch/unsealed/50MiB
                        time:   [26.411 ms 26.472 ms 26.531 ms]
                        thrpt:  [1.8404 GiB/s 1.8445 GiB/s 1.8488 GiB/s]
                 change:
                        time:   [-2.2378% -1.9948% -1.7678%] (p = 0.00 < 0.05)
                        thrpt:  [+1.7997% +2.0354% +2.2890%]
                        Performance has improved.
writer/encode_batch/trace/50MiB
                        time:   [25.886 ms 25.897 ms 25.908 ms]
                        thrpt:  [1.8847 GiB/s 1.8855 GiB/s 1.8863 GiB/s]
                 change:
                        time:   [-4.0088% -3.9115% -3.8162%] (p = 0.00 < 0.05)
                        thrpt:  [+3.9676% +4.0708% +4.1762%]
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Benchmarking writer/blob_cache_set_unsealed_batch/file_unsorted/50MiB: Warming up for 1.0000 s
Warning: Unable to complete 10 samples in 1.0s. You may wish to increase target time to 3.2s.
writer/blob_cache_set_unsealed_batch/file_unsorted/50MiB
                        time:   [449.27 ms 449.99 ms 450.83 ms]
                        thrpt:  [110.91 MiB/s 111.11 MiB/s 111.29 MiB/s]
                 change:
                        time:   [-0.0738% +0.1045% +0.2881%] (p = 0.31 > 0.05)
                        thrpt:  [-0.2872% -0.1044% +0.0738%]
                        No change in performance detected.
Benchmarking writer/blob_cache_set_unsealed_batch/mem_unsorted/50MiB: Warming up for 1.0000 s
Warning: Unable to complete 10 samples in 1.0s. You may wish to increase target time to 1.7s or enable flat sampling.
writer/blob_cache_set_unsealed_batch/mem_unsorted/50MiB
                        time:   [44.313 ms 45.284 ms 46.070 ms]
                        thrpt:  [1.0599 GiB/s 1.0783 GiB/s 1.1019 GiB/s]
                 change:
                        time:   [-16.946% -8.0952% -1.1480%] (p = 0.05 < 0.05)
                        thrpt:  [+1.1613% +8.8083% +20.404%]
                        Performance has improved.
Found 2 outliers among 10 measurements (20.00%)
  2 (20.00%) low severe

Benchmarking writer/indexed_drain/mem_sorted/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 8.1s or enable flat sampling.
writer/indexed_drain/mem_sorted/50MiB
                        time:   [132.69 ms 137.28 ms 139.78 ms]
                        thrpt:  [357.72 MiB/s 364.21 MiB/s 376.81 MiB/s]
                 change:
                        time:   [-7.2565% +0.6187% +8.7674%] (p = 0.89 > 0.05)
                        thrpt:  [-8.0607% -0.6149% +7.8243%]
                        No change in performance detected.
Benchmarking writer/indexed_drain/mem_unsorted/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 9.5s or enable flat sampling.
writer/indexed_drain/mem_unsorted/50MiB
                        time:   [128.12 ms 134.04 ms 136.82 ms]
                        thrpt:  [365.44 MiB/s 373.01 MiB/s 390.26 MiB/s]
                 change:
                        time:   [-5.8074% -2.1798% +2.1732%] (p = 0.34 > 0.05)
                        thrpt:  [-2.1270% +2.2284% +6.1654%]
                        No change in performance detected.
writer/indexed_drain/file_sorted/50MiB
                        time:   [395.78 ms 396.06 ms 396.34 ms]
                        thrpt:  [126.16 MiB/s 126.24 MiB/s 126.33 MiB/s]
                 change:
                        time:   [-0.0686% +0.0223% +0.1199%] (p = 0.65 > 0.05)
                        thrpt:  [-0.1198% -0.0223% +0.0686%]
                        No change in performance detected.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) low mild
  1 (10.00%) high mild
writer/indexed_drain/file_unsorted/50MiB
                        time:   [350.85 ms 351.11 ms 351.38 ms]
                        thrpt:  [142.29 MiB/s 142.41 MiB/s 142.51 MiB/s]
                 change:
                        time:   [-0.0322% +0.0749% +0.1824%] (p = 0.20 > 0.05)
                        thrpt:  [-0.1821% -0.0748% +0.0322%]
                        No change in performance detected.
```

ruchirK commented:
Feature benchmarks show some surprising regressions -- rerunning to see what among these might be noise

```
NAME                      |    THIS     |    OTHER    |  Regression?  | 'THIS' is:
----------------------------------------------------------------------------------------------------
FastPathFilterNoIndex     |       0.615 |       0.523 |    !!YES!!    | 17.6 pct   slower
FastPathFilterIndex       |       0.176 |       0.175 |      no       | 0.6 pct   slower
FastPathOrderByLimit      |       0.291 |       0.299 |      no       | 2.7 pct   faster
Insert                    |       2.777 |       2.690 |      no       | 3.2 pct   slower
Update                    |       1.991 |       1.926 |      no       | 3.4 pct   slower
InsertAndSelect           |       3.605 |       3.462 |      no       | 4.1 pct   slower
OrderBy                   |       7.095 |       7.006 |      no       | 1.3 pct   slower
CountDistinct             |       0.989 |       0.997 |      no       | 0.8 pct   faster
MinMax                    |       2.846 |       2.857 |      no       | 0.4 pct   faster
GroupBy                   |       8.446 |       8.309 |      no       | 1.6 pct   slower
CrossJoin                 |       2.087 |       2.083 |      no       | 0.2 pct   slower
Retraction                |       2.374 |       2.285 |      no       | 3.9 pct   slower
CreateIndex               |       0.196 |       0.189 |      no       | 3.9 pct   slower
DeltaJoin                 |       0.359 |       0.361 |      no       | 0.7 pct   faster
DifferentialJoin          |       0.669 |       0.663 |      no       | 0.8 pct   slower
FinishOrderByLimit        |       0.142 |       0.148 |      no       | 4.4 pct   faster
KafkaRaw                  |       1.184 |       1.205 |      no       | 1.8 pct   faster
KafkaEnvelopeNoneBytes    |       5.379 |       3.350 |    !!YES!!    | 60.6 pct   slower
KafkaUpsert               |       1.999 |       2.052 |      no       | 2.6 pct   faster
KafkaUpsertUnique         |       3.308 |       2.436 |    !!YES!!    | 35.8 pct   slower
KafkaRecovery             |      21.537 |      20.836 |      no       | 3.4 pct   slower
ExactlyOnce               |       5.422 |       6.101 |      no       | 11.1 pct   faster
PgCdcInitialLoad          |       2.279 |       2.172 |      no       | 4.9 pct   slower
PgCdcStreaming            |       1.917 |       1.911 |      no       | 0.3 pct   slower
QueryLatency              |       1.334 |       1.364 |      no       | 2.2 pct   faster
ConnectionLatency         |       0.814 |       0.800 |      no       | 1.8 pct   slower
```

danhhz commented:
Is it possible this is the O(n^2) memcpys I asked about? I realized that, since we moved compaction to the background, the only microbenchmark that covers it is now is end_to_end/load. And that one only has an s3 variant (btw, all you have to do to run these on a scratch machine is `source src/persist/s3_test_env_mz.sh` and it will pick them up).

It's probably worth adding a microbenchmark specifically for `Maintainer::compact_trace_blocking`. I'm thinking something simple that doesn't worry about coverage of compaction or consolidation:

- Construct the two input batches by alternating batches from DataGeneratorBatchIter, one to batch0, one to batch1, then repeat until the batch iter is exhausted
- Save the input batches in blob and create a CompactTraceReq
- As the benchmark loop: call Maintainer::compact_trace_blocking

My guess is this will help triage the feature benchmark regressions. Don't forget the ability to override MZ_PERSIST_RECORD_COUNT when running the benchmark to check various sizes

ruchirK commented:
Ok this is ready for a look! The first commit has the new compact trace benchmark + a detailed analysis of how they look currently in three different size + num batch configurations, the second commit has the new/modified invariants for trace batches and the third commit actually handles compacting part-by-part with bounded overhead (now without quadratic work :)), and new tests + an evaluation of the compaction microbenchmarks.

I owe the reviewers a full set of microbenchmark results + feature benchmark results + an analysis of how this affects the open loop tests at high load. I'm going to get to all of those first thing tomorrow AM but as a teaser, the snapshot benchmark results now look to be identical to main. I'm not sure if that was affected by the quadratic-ness fix, or if that was just noise, earlier, but these results seem very repeatable.

```
MZ_PERSIST_RECORD_COUNT=819200 MZ_PERSIST_RECORD_SIZE_BYTES=64 MZ_PERSIST_BATCH_MAX_COUNT=131072
snapshot/blob_get/mem/50MiB
                        time:   [20.445 ms 20.455 ms 20.466 ms]
                        thrpt:  [2.3858 GiB/s 2.3871 GiB/s 2.3882 GiB/s]
                 change:
                        time:   [+1.0717% +1.1475% +1.2212%] (p = 0.00 < 0.05)
                        thrpt:  [-1.2065% -1.1345% -1.0603%]
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

Benchmarking snapshot/mem/unsealed/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 12.8s, or reduce sample count to 30.
snapshot/mem/unsealed/50MiB
                        time:   [128.50 ms 129.55 ms 130.89 ms]
                        thrpt:  [382.00 MiB/s 385.96 MiB/s 389.12 MiB/s]
                 change:
                        time:   [-1.0246% -0.2110% +0.7822%] (p = 0.69 > 0.05)
                        thrpt:  [-0.7761% +0.2114% +1.0352%]
                        No change in performance detected.
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high severe
Benchmarking snapshot/mem/trace/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 15.6s, or reduce sample count to 30.
snapshot/mem/trace/50MiB
                        time:   [148.77 ms 148.89 ms 149.02 ms]
                        thrpt:  [335.53 MiB/s 335.82 MiB/s 336.09 MiB/s]
                 change:
                        time:   [-1.1674% -1.0590% -0.9528%] (p = 0.00 < 0.05)
                        thrpt:  [+0.9619% +1.0703% +1.1812%]
                        Change within noise threshold.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

Benchmarking snapshot/file/unsealed/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 12.9s, or reduce sample count to 30.
snapshot/file/unsealed/50MiB
                        time:   [129.01 ms 129.10 ms 129.19 ms]
                        thrpt:  [387.03 MiB/s 387.31 MiB/s 387.58 MiB/s]
                 change:
                        time:   [-0.8731% -0.7698% -0.6739%] (p = 0.00 < 0.05)
                        thrpt:  [+0.6785% +0.7758% +0.8808%]
                        Change within noise threshold.
Benchmarking snapshot/file/trace/50MiB: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 16.3s, or reduce sample count to 30.
snapshot/file/trace/50MiB
                        time:   [151.28 ms 151.34 ms 151.40 ms]
                        thrpt:  [330.25 MiB/s 330.39 MiB/s 330.52 MiB/s]
                 change:
                        time:   [+3.1778% +3.3329% +3.4791%] (p = 0.00 < 0.05)
                        thrpt:  [-3.3621% -3.2254% -3.0799%]
                        Performance has regressed.
```

danhhz commented on src/persist/src/indexed/background.rs:
```diff
@@ -164,69 +192,162 @@ impl<B: Blob> Maintainer<B> {
-        for ((_, _), ts, _) in updates.iter_mut() {
-            ts.advance_by(desc.since().borrow());
-        }
+            // Drain all of the records <= threshold out of the various holding
```
one thing that tripped me up while reviewing was trying to get a feel for the size of the output batches under various worst-case conditions. specifically, I'm not sure I see anything that puts a lower bound on the size of the parts this outputs. so the situation I'm thinking out is as batches get repeatedly compacted over time, whether anything keeps them from being chopped up in increasingly more and more parts.

an alternative is instead of flushing `updates` every time we finish one of the input batches, we repeatedly pull whichever one is necessary until `updates` hits some certain size (not just a num records, it would include key and val lens). the "what is the maximum memory usage" analysis in this case is a bit more complicated, but I _think_ it ends up as something like `first batch part size + second batch part size + 2*key_val_data_max_len`?

ruchirK replied:
So, I think there's a slight misunderstanding here. Zooming out a little bit:

`first_updates` and `second_updates` are buffers holding data from one part of the first/second trace batch respectively. We know we can stop pulling data when both are empty, and they are each bounded by the size of a `BlobTraceBatchPart`

(poorly named) `updates` is the buffer that holds the subset of `first_updates` and `second_updates` that can be forwarded to `since`, consolidated, and then placed into `builder`. We move records from `updates` -> `builder` one by one after they have been consolidated.

As we push things into builder, we only write a new `BlobTraceBatchPart` into storage once a `ColumnarRecords` has been filled, and also once at the very end in case there's leftover data that did not fill up a `ColumnarRecords`. So the lower bound on the number of batches is `(size of consolidated data) / (size limits of a ColumnarRecords) + 1` or something like that.

Does that make sense? I think we are already doing what you've suggested here - we repeatedly

1. figure out which of first_updates and second_updates is empty, and attempt to read the next batch
2. figure out the threshold of updates that can be consolidated
3. consolidate them and move them into builder
4. write out batches as the number of records in builder fills up a ColumnarRecords
5. rinse and repeat.

Let me know if this explanation isn't clear.

danhhz replied:
let's hop on a zoom! we're definitely talking past each other, but I don't yet see exactly how

danhhz replied:
talked about this offline. i was misunderstanding exactly what the code was doing, so this concern is a no-op. new feedback is that we might want to think a bit about how to tweak the code to make what is going on a bit more obvious

ruchirK replied:
I added a new commit that adds 3 helper functions to make what the compaction loop is doing hopefully a bit more obvious

danhhz commented on src/persist/src/indexed/columnar.rs:
```diff
@@ -511,8 +511,10 @@ impl Default for ColumnarRecordsVecBuilder {
+    /// Create a new ColumnarRecordsVecBuilder with a specified max size.
+    ///
+    /// Only used for testing.
+    pub(crate) fn new_with_len(key_val_data_max_len: usize) -> Self {
```
why'd we lose the #[cfg(test)]?

ruchirK replied:
we need to use this function in `compact_trace_blocking` in the case when the `key_val_max_len` passed to `Maintainer` is `Some(_)`. `compact_trace_blocking` is not cfg(test), and so the compiler complains if this function is, even though the key_val_max_len is only `Some(_)` in `Maintainer::new_for_testing` which is `cfg(test)` but the compiler can't infer that.

ruchirK replied:
I couldn't think of any nicer workarounds, curious if you have any ideas!

danhhz reviewed (commented):
first batch of comments! first two commits look basically ready to go. I had some questions on the strategy we're using to walk through the input batches, so I held off on line level review on the impl

danhhz commented:
> As expected, mem and file blobs take
~10x as much time in this setup compared to the default, however s3 blobs take
only about ~5x as much time (the aggregate throughput doubles). It's unclear whether
this is a product of s3 latencies being inherently noisy, or something else.
>
> My unsubstantiated hypothesis is that perhaps at
the smaller sizes s3 blobs are not able to take advantage of splitting a BlobTraceBatchPart
into evenly sized chunks for the s3 uploads (or perhaps there's one large chunk and one tiny
chunk at the end?).

yeah, this is probably the s3 multipart stuff chunking things at 8MiB

> We don't do the real linear merge as compaction requires us to both forward times
and consolidate multiple updates at the forwarded times. I believe that doing so
is possible in linear time, but doing so in a single pass was complicated enough
that I am leaving it for future work.

agree and agree

aljoscha commented:
I have read this and think it intuitively makes sense. I don't have any useful comments/suggestions, though.

ruchirK commented on src/persist/src/indexed/background.rs:
```diff
@@ -243,6 +257,21 @@ mod tests {
         )
     }
 
+    fn blob_trace_batch_part(
```
I can't tell if this is a statement of a mild preference or if you'd like me to remove the helper and leave it the way it was before - i agree it doesn't do a lot and am happy to remove it just lmk

danhhz replied:
it's both! :)

ruchirK commented on src/persist/src/indexed/arrangement.rs:
```diff
@@ -407,23 +407,34 @@ impl Arrangement {
-            level: 0,
-            size_bytes,
+
+        let drained = if !updates.is_empty() {
```
done. I added this change + the other one directly to the second commit because they were so small

ruchirK commented on src/persist/src/indexed/background.rs:
```diff
@@ -96,6 +99,22 @@ impl<B: BlobRead> Maintainer<B> {
+    }
+
+    /// Returns a new [Maintainer].
+    pub fn new_for_testing(
```
no good reason - fixed!

ruchirK commented on src/persist/src/indexed/background.rs:
```diff
@@ -164,69 +192,162 @@ impl<B: Blob> Maintainer<B> {
-                index: 0,
-                updates,
-            };
+            i = 0;
```
done.

ruchirK commented:
feature benchmarks look good too! (with some rather surprising speedups actually)

```
NAME                      |    THIS     |    OTHER    |  Regression?  | 'THIS' is:
----------------------------------------------------------------------------------------------------
FastPathFilterNoIndex     |       0.532 |       0.543 |      no       | 2.0 pct   faster
FastPathFilterIndex       |       0.180 |       0.180 |      no       | 0.3 pct   faster
FastPathOrderByLimit      |       0.297 |       0.284 |      no       | 4.8 pct   slower
Insert                    |       2.770 |       2.787 |      no       | 0.6 pct   faster
Update                    |       1.958 |       1.973 |      no       | 0.8 pct   faster
InsertAndSelect           |       3.562 |       3.454 |      no       | 3.1 pct   slower
OrderBy                   |       7.169 |       7.030 |      no       | 2.0 pct   slower
CountDistinct             |       1.027 |       0.998 |      no       | 2.9 pct   slower
MinMax                    |       2.873 |       2.862 |      no       | 0.4 pct   slower
GroupBy                   |       8.470 |       8.313 |      no       | 1.9 pct   slower
CrossJoin                 |       2.119 |       2.105 |      no       | 0.7 pct   slower
Retraction                |       2.271 |       2.284 |      no       | 0.6 pct   faster
CreateIndex               |       0.204 |       0.192 |      no       | 6.3 pct   slower
DeltaJoin                 |       0.365 |       0.361 |      no       | 1.3 pct   slower
DifferentialJoin          |       0.671 |       0.675 |      no       | 0.6 pct   faster
FinishOrderByLimit        |       0.139 |       0.149 |      no       | 6.8 pct   faster
KafkaRaw                  |       1.179 |       1.183 |      no       | 0.4 pct   faster
KafkaEnvelopeNoneBytes    |       3.383 |       3.360 |      no       | 0.7 pct   slower
KafkaUpsert               |       2.363 |       2.927 |      no       | 19.3 pct   faster
KafkaUpsertUnique         |       2.300 |       2.541 |      no       | 9.5 pct   faster
KafkaRecovery             |      20.377 |      21.619 |      no       | 5.7 pct   faster
ExactlyOnce               |       4.576 |       5.859 |      no       | 21.9 pct   faster
PgCdcInitialLoad          |       2.438 |       2.226 |      no       | 9.5 pct   slower
PgCdcStreaming            |       1.928 |       1.944 |      no       | 0.8 pct   faster
QueryLatency              |       1.342 |       1.367 |      no       | 1.9 pct   faster
ConnectionLatency         |       0.832 |       0.814 |      no       | 2.2 pct   slower
==> Test case workflow-default succeeded
```

I'll fold these and the full microbenchmark results into the commit message before merging

danhhz reviewed (approved):
LGTM

ruchirK commented:
TFTR! updated the commit messages, removed the unhelpful test wrapper and am merging on green!

ruchirK commented:
I'm not immediately sure what's failing in the upgrade test, but it seems to be doing so repeatedly

ruchirK commented:
oh whoops i was looking in the totally wrong part of the test output, i know what this is

ruchirK commented:
woohoo! after thinking about it a bit it was really straightforward to extract out the new invariant that trace batch parts must be nonempty. In the second commit, that amounted to ripping out a few lines of validation code and flipping one test case from an error to an ok. In the last commit, it turns into "keep fetching trace batch parts when the input buffer is empty until either you run out of batch parts to fetch, or the buffer is no longer empty"

```
             // Pull more data if either of the holding pens from first or second are empty
-            if first_updates.is_empty() {
-                if let Some(key) = first_batch_iter.next() {
-                    Self::load_trace_batch_part(&blob, key, &mut first_updates)?;
+            //
+            // TODO: we can simplify this logic once empty trace batch parts are disallowed.
+            while first_updates.is_empty() {
+                match first_batch_iter.next() {
+                    Some(key) => Self::load_trace_batch_part(&blob, key, &mut first_updates)?,
+                    None => break,
                 }
             }

-            if second_updates.is_empty() {
-                if let Some(key) = second_batch_iter.next() {
-                    Self::load_trace_batch_part(&blob, key, &mut second_updates)?;
+            while second_updates.is_empty() {
+                match second_batch_iter.next() {
+                    Some(key) => Self::load_trace_batch_part(&blob, key, &mut second_updates)?,
+                    None => break,
                 }
             }
```

You have to make this be a `while` loop instead of an `if` statement now because you can't always guarantee that a trace batch part will give you some data. Added several test cases to check what happens if empty trace batches are interleaved with nonempty ones or if a trace batch is fully empty (with many empty parts or with no parts).

TFTR everyone for the detailed review, I'm going to merge this one on green and open up a followup pr to ban empty trace batch parts that we can merge later
